### DISCUSSION TITLE
Edge Impulse - 'multiple definition of' errors

### DISCUSSION THREAD

**Discussion by OOHehir** [LIFsCode/ELOC-3.0#29] [General]
@EDsteve @LIFsCode 

Hello all, 

I've spent hours wrestling trying to get the Edge Impulse code moved to external files but keeping running into repeated issue about errors 
_multiple definition of ...._ 
I attach a sample output below. Perhaps I'm missing something obvious (quite likely) or perhaps it's something to do with the way the Edge Impulse code is structured.  Either way perhaps you've come across something similar in the project before?

You'll probably need to see the source code, I'll format it correctly (hopefully tomorrow) & upload it to the air branch if you get a chance to look at it..

```
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `ei::spectral::processing::spectral_power_edges(ei::ei_matrix*, ei::ei_matrix*, ei::ei_matrix*, ei::ei_matrix*, float)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/dsp/spectral/processing.hpp:318: multiple definition of `ei::spectral::processing::spectral_power_edges(ei::ei_matrix*, ei::ei_matrix*, ei::ei_matrix*, ei::ei_matrix*, float)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/dsp/spectral/processing.hpp:318: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `ei::spectral::processing::periodogram(ei::ei_matrix*, ei::ei_matrix*, ei::ei_matrix*, float, unsigned short)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/dsp/spectral/processing.hpp:376: multiple definition of `ei::spectral::processing::periodogram(ei::ei_matrix*, ei::ei_matrix*, ei::ei_matrix*, float, unsigned short)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/dsp/spectral/processing.hpp:376: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `fill_input_tensor_from_matrix(ei::ei_matrix*, TfLiteTensor*)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_helper.h:60: multiple definition of `fill_input_tensor_from_matrix(ei::ei_matrix*, TfLiteTensor*)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_helper.h:60: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `fill_input_tensor_from_signal(ei::ei_signal_t*, TfLiteTensor*)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_helper.h:114: multiple definition of `fill_input_tensor_from_signal(ei::ei_signal_t*, TfLiteTensor*)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_helper.h:114: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `fill_output_matrix_from_tensor(TfLiteTensor*, ei::ei_matrix*)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_helper.h:237: multiple definition of `fill_output_matrix_from_tensor(TfLiteTensor*, ei::ei_matrix*)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_helper.h:237: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `fill_result_struct_from_output_tensor_tflite(ei_impulse const*, TfLiteTensor*, TfLiteTensor*, TfLiteTensor*, ei_impulse_result_t*, bool)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_helper.h:293: multiple definition of `fill_result_struct_from_output_tensor_tflite(ei_impulse const*, TfLiteTensor*, TfLiteTensor*, TfLiteTensor*, ei_impulse_result_t*, bool)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_helper.h:293: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_nn_inference_from_dsp(ei_nn_tflite_eon_t const*, ei::ei_signal_t*, ei::ei_matrix*)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_eon.h:174: multiple definition of `run_nn_inference_from_dsp(ei_nn_tflite_eon_t const*, ei::ei_signal_t*, ei::ei_matrix*)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_eon.h:174: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_nn_inference(ei_impulse const*, ei::ei_matrix*, ei_impulse_result_t*, bool)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_eon.h:232: multiple definition of `run_nn_inference(ei_impulse const*, ei::ei_matrix*, ei_impulse_result_t*, bool)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_eon.h:232: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/model-parameters/dsp_blocks.h:35: multiple definition of `ei_dsp_blocks'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/model-parameters/dsp_blocks.h:35: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_nn_inference_image_quantized(ei_impulse const*, ei::ei_signal_t*, ei_impulse_result_t*, bool)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_eon.h:285: multiple definition of `run_nn_inference_image_quantized(ei_impulse const*, ei::ei_signal_t*, ei_impulse_result_t*, bool)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_eon.h:285: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `extract_tflite_eon_features(ei::ei_signal_t*, ei::ei_matrix*, void*, float)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_eon.h:361: multiple definition of `extract_tflite_eon_features(ei::ei_signal_t*, ei::ei_matrix*, void*, float)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/inferencing_engines/tflite_eon.h:361: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_inference':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:128: multiple definition of `run_inference'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:128: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `process_impulse_continuous':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:251: multiple definition of `process_impulse_continuous'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:251: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/model-parameters/model_variables.h:29: multiple definition of `ei_classifier_inferencing_categories'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/model-parameters/model_variables.h:29: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `ei_construct_impulse':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:428: multiple definition of `ei_construct_impulse'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:428: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_classifier_image_quantized':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:594: multiple definition of `run_classifier_image_quantized'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:594: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `process_impulse':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:166: multiple definition of `process_impulse'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:166: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_classifier_init':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:617: multiple definition of `run_classifier_init'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:617: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_classifier_deinit':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:658: multiple definition of `run_classifier_deinit'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:658: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_classifier_continuous':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:679: multiple definition of `run_classifier_continuous'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:679: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `run_classifier':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:720: multiple definition of `run_classifier'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_run_classifier.h:720: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `ei_classifier_smooth_init(ei_classifier_smooth*, unsigned int, unsigned char, float, float)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_classifier_smooth.h:48: multiple definition of `ei_classifier_smooth_init(ei_classifier_smooth*, unsigned int, unsigned char, float, float)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_classifier_smooth.h:48: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `ei_classifier_smooth_update(ei_classifier_smooth*, ei_impulse_result_t*)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_classifier_smooth.h:66: multiple definition of `ei_classifier_smooth_update(ei_classifier_smooth*, ei_impulse_result_t*)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_classifier_smooth.h:66: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o: in function `ei_classifier_smooth_free(ei_classifier_smooth*)':
/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_classifier_smooth.h:135: multiple definition of `ei_classifier_smooth_free(ei_classifier_smooth*)'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/edge-impulse-sdk/classifier/ei_classifier_smooth.h:135: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/model-parameters/model_variables.h:33: multiple definition of `ei_dsp_config_3'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/model-parameters/model_variables.h:33: first defined here
/home/owen/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev-ei/src/main.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/model-parameters/model_variables.h:31: multiple definition of `ei_dsp_config_3_axes'; .pio/build/esp32dev-ei/src/EdgeImpulse.o:/home/projects/ELOC-3.0/eloc610LowPowerPartition/src/model-parameters/model_variables.h:31: first defined here
```

**Comment by LIFsCode** [comment]
I'm just guessing as I don't see it in the source code, but from what I saw in your previous commit to the AI branche, I would suspect some sort of C & C++ name mangling.

Did you correctly use "extern "C" {}" wherever needed? 

I noticed that is not part of the generated inferencing header files, maybe that's the problem.

**Comment by OOHehir** [comment]
Many thanks, I think that could be it!

**Comment by OOHehir** [comment]
Turns out simply moving the #include of the edge impulse header file solved the issue.

**Comment by LIFsCode** [comment]
ok, so I'm pretty sure this is a C / C++ name mangling issue.

Actually even if moving the include solves this issue currenlty it is very likely to reoocur once any of the includes are modified.
Better solution would be to strictly place an `extern "C" {}` arround or within all pure C includes.

**Comment by OOHehir** [comment]
I agree this is going to be a finicky issue, but the issue is that the Edge Impulse code is built on other code, particularly TensorFlow, all told there's about 1000 files in total. Modifying them would obviously be a significant undertaking. 

There is a possibility that the Edge Impulse library code may have to be updated when a new model is created. Additionally a more open source solution (a custom TensorFlow model) may, in time, replace the Edge Impulse solution.

**Reply by LIFsCode** [reply]
Totally agree, modifying generated files does not make sense. I had something in mind like, adding a kind of "wrapper" header file, which includes all the required EI  interface headers and the rest of the program is only including this header.

e.g. 

edge_impulse_inc.h

```
#pragma once

#ifdef __cplusplus
extern "C" {
#endif


#ifdef EDGE_IMPULSE_ENABLED
    #include <EdgeImpulse.hpp>
    #include <test_samples.h>   // TODO: Move this to unit test code
   // #include "trumpet_inferencing.h"
    extern inference_t inference;
    extern const uint32_t sample_buffer_size;
    extern signed short sampleBuffer[];
    extern bool debug_nn;
    extern int print_results;
    extern bool record_status;
   // add all includes here which are required for interfacing edge impulse
#endif

#ifdef __cplusplus
}
#endif
```

and in main.cpp (or in any other ELOC code

`#include edge_impulse_inc.h`

This should keep everything cleen without the need for special include orders or modifications of generated code

**Reply by OOHehir** [reply]
@LIFsCode Thanks for the suggestion, I'm fairly certain I went down this road it didn't fix the issue. I'll try once again to be certain & perhaps send you the commit just I case I missed something..

**Reply by OOHehir** [reply]
@LIFsCode Well, I gave this another go. A few issues crop up:

1. The most significant issue is that the main header file "trumpet_inferencing.h" is a C++ file but seems to pull in other C so putting guards around it doesn't work. 
2. There are structures, (eg:  typedef struct ei_signal_t* signal &  typedef ei_impulse_result_t *result) that contain references to function callbacks. Trying to instantiate these in a class constructor leads to errors, it seem they need to be created in a global context. This could be because the Edge Impulse code changes dynamically at build time depending on the target system.
4. I created a C++ class (EdgeImpulse.cpp) that includes "trumpet_inferencing.h". In order to create the 'signal' & 'result' structures, as referenced above, in main.cpp I need to include the same header. Once that happens the multiple definition errors occur.

Perhaps you can spot an obvious error or perhaps a solution, if you can I would be grateful!

You can find the code as it stand on the branch [ai-deadend](https://github.com/LIFsCode/ELOC-3.0/tree/ai-deadend)

Regards,

Owen

**Comment by EDsteve** [comment]
@OOHehir 
Great to hear you got it working (at least for now) :)
The goal indeed is to use a custom solution without EI. But we should have something ready to test in the jungle this year. Custom model or EI. Whatever wins the race. Is EI kind of working now? Let me know when you have something i can play around with. Excited :)

**Reply by OOHehir** [reply]
@LIFsCode @EDsteve Over the next few days I'll work on configuring, starting & stopping the I2S mic from external of the EI code. From that point it should be more straightforward to integrate running the Edge Impulse model & saving sound to file simultaneously. @LIFsCode you have any thoughts/ preference on the code to setup, start & stop the I2S mic? I think you expressed an interest in moving it to separate files.

@EDsteve Do you think you'll have an improved set of samples to construct a model with? I'm still using the Rumble_Detector model, which isn't very accurate.

**Comment by EDsteve** [comment]
@OOHehir 
If i understand you correctly it will be difficult to change the EI model easily? Because i am sure it needs to be changed frequently in the beginning. And it would be great if i would be able to do it myself. 

I will send you a seperate email about the trimmed files. But yes. There is progress.

Side note: Not so sure if you have looked at the ELOC app yet. Just so you know the already included functionality. It's only available at [apkpure ](https://m.apkpure.com/eloc-listener-control-panel/de.eloc.eloc_control_panel) for now because i am moving the app to a different PlayStore account. Or watch the [outdated video](https://www.youtube.com/watch?v=Jb5j_fkpN6I&t=7s) i made for our students to get an impression.

**Reply by OOHehir** [reply]
@EDsteve No, changing the model should be quite painless, just changing some files (in lib/edge_impulse).

I'll take a look app when I get a chance

**Comment by OOHehir** [comment]
@LIFsCode @EDsteve More issues with this file setup.. The Edge Impulse dynamically creates structures to run the inference. These can't be created in main.cpp as I need the header for the structure & then run into the issues above. Creating them in the library header (EdgeImpuse.hpp) results in the same issue. 

I think I'll have to abandon this file re-arrangement for the moment, it may be a dead-end & I'd like to get some sort of file logging working to capture any outputs from the EI inference process so it can be tested.

Perhaps at some point in the future a Edge Impulse may change the structure of the library to make more straightforward to include without these issues..

**Reply by LIFsCode** [reply]
@OOHehir  I will have a look at this once I finde some time.

I think in the meantime it's suitable as it is we can rearrange and clean up the code later.

**Comment by LIFsCode** [comment]
@OOHehir hi, sorry I forgot a little bit about this issue. 
I just head a look at the ai-deadend branch an I think I solved the problems with the multiple definitions: see comit c81a46d

I was wrong about the assumption with the "extern C" header includes, the problem was within "procesing.hpp" which  contains implementation of various non inline non static functions.  This caused multiple definition errors if processing.hpp is included from  more than 1 .cpp file

Actually you have already been on the correct path, I noticed some comments in the code from you about trumpet_inferencing.h causing this errors.
I simple removed the include of trumpet_inferencing.h from the main.cpp. It is actually not needed as all required interfaces are defined elsewhere.

Feel free to have a look at the current ai-daedend branch. I think thus it should easily be possible to separate the AI stuff.

**Reply by OOHehir** [reply]
@LIFsCode Thanks for looking at the code. 

I believe I tried this but ran into further issues. I think the reason that its compiling for you at the moment is that I've commented out all the EI code (starting around [here](https://github.com/LIFsCode/ELOC-3.0/blob/c81a46d190d4cbcddb7854a2e20ef6774146da96/eloc610LowPowerPartition/src/main.cpp#L860)). Apologies for the false hope but the reason I commented it all out was to make the code as simple as possible to try & get a successful compile.

trumpet_inferencing.h is required for structures (eg: typedef struct ei_signal_t* signal & typedef ei_impulse_result_t *result). I tried declaring these in EdgeImpulse.cpp but ran into the issues as mentioned [here](https://github.com/LIFsCode/ELOC-3.0/discussions/29#discussioncomment-7134004).

If you have time you may wish to investigate further & see if you can find a solution but I've lost quite a few hours of my life so far in the process!

**Comment by LIFsCode** [comment]
ok, sorry I didn't expect the commented out sections to be runable, as they don't even implent the EdgeImpulse class.

Actually I'm pretty confused, because these sections still contain the run_classifier, which I assume to be the main inference logic. I would have expected this to be somehow encapsulated within a different file.

Anyway I just enabled the commented sections and fix all compile errors. Note it is not yet clean code, but since I'm not sure about your concept for it I didn't want to refactor too much.

Basic stuff I did (8ffc089):

- added missing include files (only the specific ones which are needed, not the whole trumpet_inferencing.h
- used correct ei namespace for signal_t
- used correct class instance references for your EdgeImpulse class
- added a wrapper for microphone_audio_signal_get_data(). the signal_t class requires a non class member function. Since you added microphone_audio_signal_get_data to EdgeImpulse class, a non member wrapper is required to assign it to signal_t
- I moved print_results out of the EdgeImpulse class. So far I don't see any class specific behavior in this variable. but it could be accessed as member variable as well
- I moved run_classifier_continuous() to EdgeImuplse class. Adding the required includes here would include nearly all the EI stuff, which also includes the inproper header files with non-static non-inline function definitions (which I still consider a bug within the edgeImpulse generation framework)

I uncommented another section (6707353), which I assume is not in use, so potentially you could ignore this commit, as it would also make the inference struct accessible outside the class. I'm not sure if that makes sense.

After all I must admit, I did all these changes without thinking about the functionality, I was only fixing the compile errors.

**Comment by OOHehir** [comment]
@LIFsCode Thanks for taking a look. I tried a couple of approaches to split off the EI code into separate files & hence why it looks messy. I'm in the process of moving the sound capturing into the I2MEMSampler & populating the EI & WAV buffers from there. Once that's complete I'll come back to this & look at your suggestions. 

One of my aims was to try & keep the process of changing the EI models (by replacing all the files in the lib/edge-impluse folder) as straightforward as possible. Including the single header file 'trumpet_inferencing.h' was guaranteed to accomplish this but perhaps by carefully picking 'nested' header files (such as "edge-impulse-sdk/classifier/ei_classifier_types.h") as you have suggested this can also be accomplished.

**Comment by OOHehir** [comment]
@LIFsCode Thanks again for looking at this. I've tidied up all the commented stuff a bit further. 

I previously spent quite a bit of time trying to get
```
ei::signal_t 
ei_impulse_result_t
```
Into the EdgeImpulse class & from there I planned to encapsulate nearly all the inference material. With that in place I hoped to make the inference process into a thread (similar to the wav_writer). I think with the two types above remaining in main.cpp it'll still be possible to have a thread but the code for it will have to remain in main.cpp.

@EDsteve 
I've just tidied up the ai-deadend branch (it moves some of the code into a separate class) with help from @LIFsCode & compared the test results of it against the working ai branch:
```
ai-deadend branch
I (9983) main: Test model predictions:
I (9984) main:     (DSP: 1349 ms., Classification: 44 ms., Anomaly: 0 ms.)
I (9984) main:     background: 0.207031
I (9989) main:     trumpet: 0.792969
W (9993) main: Test of trumpet sample appears to be poor, check model!
I (10000) main: Running test category: other
I (10304) main: Test model predictions:
I (10305) main:     (DSP: 245 ms., Classification: 44 ms., Anomaly: 0 ms.)
I (10306) main:     background: 0.996094
I (10311) main:     trumpet: 0.000000

ai branch
I (9993) main: Test model predictions:
I (9994) main:     (DSP: 1364 ms., Classification: 44 ms., Anomaly: 0 ms.)
I (9994) main:     background: 0.207031
I (9999) main:     trumpet: 0.792969
W (10003) main: Test of trumpet sample appears to be poor, check model!
I (10010) main: Running test category: other
I (10314) main: Test model predictions:
I (10315) main:     (DSP: 245 ms., Classification: 44 ms., Anomaly: 0 ms.)
I (10316) main:     background: 0.996094
I (10321) main:     trumpet: 0.000000`
```

As you can see the predictions are identical (except for DSP time, which is other process dependent) so I think we can be reasonably certain that the functionality is identical. I plan to merge these branches & from that they can be merged into master/ main.

**Comment by OOHehir** [comment]
Resolved
